plonky2/src/hash/merkle_tree.rs

```diff
@@ -83,13 +83,21 @@ pub(crate) fn capacity_up_to_mut<T>(v: &mut Vec<T>, len: usize) -> &mut [MaybeUn
     }
 }
 
-pub(crate) fn fill_subtree<F: RichField, H: Hasher<F>>(
+pub(crate) fn fill_subtree<F, H, L, HF>(
+    index: usize,
     digests_buf: &mut [MaybeUninit<H::Hash>],
-    leaves: &[Vec<F>],
-) -> H::Hash {
+    leaves: &[L],
+    hash_fn: HF,
+) -> H::Hash
+where
+    F: RichField,
+    H: Hasher<F>,
+    L: Send + Sync,
+    HF: Send + Clone + Fn(usize, &L) -> H::Hash,
+{
     assert_eq!(leaves.len(), digests_buf.len() / 2 + 1);
     if digests_buf.is_empty() {
-        H::hash_or_noop(&leaves[0])
+        hash_fn(index, &leaves[0])
```

Comment:
I wonder if we can do this without mucking around with indices.

Reply:
So... we could do it by passing in some aux data which we split using the existing split code. It's a little less versatile, but should handle current use cases.